metaflow/plugins/kfp/kfp.py

```diff
@@ -406,74 +402,70 @@ def create_kfp_pipeline_from_flow_graph(self):
                 ).set_display_name(node.name)
 
-                set_resource_requirements(
-                    visited[node.name],
-                    KubeflowPipelines._get_resource_requirements(node),
+                KubeflowPipelines._set_container_settings(
+                    op, step_to_kfp_component_map[node.name]
                 )
 
                 if node.type == "foreach":
-                    with kfp.dsl.ParallelFor(
-                        visited[node.name].outputs["split_indexes"]
-                    ) as index:
-                        for step in node.out_funcs:
-                            build_kfp_dag(
-                                self.graph[step],
-                                visited[node.name].outputs["task_out_dict"],
-                                index,
-                            )
-                    # TODO: the join
-                    # visited[node.name].outputs["task_out_dict"]
+                    # Please see nested_parallelfor.ipynb for how this works
+                    with kfp.dsl.ParallelFor(op.output) as index:
+                        # build_kfp_dag() will halt when
+                        # a foreach join is reached.
+                        build_kfp_dag(
+                            self.graph[node.out_funcs[0]], passed_in_split_indexes=index
+                        )
+
+                    # Handle the ParallelFor join step, and pass in
+                    # passed_in_split_indexes of parent context
+                    build_kfp_dag(
+                        self.graph[node.matching_join], passed_in_split_indexes
+                    )
                 else:
                     for step in node.out_funcs:
-                        if step not in visited:
-                            build_kfp_dag(
-                                self.graph[step],
-                                visited[node.name].outputs["task_out_dict"],
-                            )
-
-                        visited[step].after(visited[node.name])
-
-            build_kfp_dag(self.graph["start"], context="")
+                        step_node = self.graph[step]
+                        if (
+                            step_node.type == "join"
+                            and self.graph[step_node.split_parents[-1]].type
+                            == "foreach"
+                        ):
+                            # halt with a foreach join is reached
+                            # see the ParallelFor and adjacent call to build_kfp_dag()
+                            # which handles the ParallelFor join.
+                            return
+                        else:
+                            build_kfp_dag(step_node, passed_in_split_indexes)
+
+            build_kfp_dag(self.graph["start"])
+
+            for step in self.graph.nodes:
+                node = self.graph[step]
+                for parent_step in node.in_funcs:
+                    visited[node.name].after(visited[parent_step])
```

Comment:
Mind adding in a comment about what this is doing. Obviously it's adding in the edges of the graph but looks like it's doing it in a breadth first traversal I assume. Looks like it's doing that over the set of `visited` nodes we traversed we constructed inside `build_kfp_dag` which from what I can tell is doing a depth-first traversal-ish with a special case for `join` nodes after it's handled all nodes in that branch.

Likely unrelated but how would this handle cycles?

Reply:
added comments!